File: cascadingFailuresTaskMigration_python/main/initialize.py
import heapq
from typing import List, Dict, Set
from input.task import Task
from input.robot import Robot
from input.group import Group
from .function import Function


class Initialize:
    def __init__(self):
        self.fault_p = 0.5

    def run(self, tasks: List[Task], robots: List[Robot],
            id_to_groups: Dict[int, Group], id_to_robots: Dict[int, Robot]):
        """Initialize task assignment and fault conditions."""
        self._ini_task(tasks, robots, id_to_groups, id_to_robots)
        self._ini_fault(id_to_robots, id_to_groups)

    def _ini_fault(self, id_to_robots: Dict[int, Robot], id_to_groups: Dict[int, Group]):
        """Initialize fault conditions for robots."""
        size = len(id_to_robots)
        fault_size = int(size * self.fault_p)
        if fault_size == 0:
            fault_size += 1
        step = size // fault_size

        # 0.1 represents the proportion of nodes with functional faults in the system
        for i in range(size):
            robot = id_to_robots[i]
            if i % step == 1:
                # Set functional fault - can simulate cascading failure scenario
                robot.fault_a = 1.0
                group_id = robot.group_id
                group = id_to_groups[group_id]
                group.group_capacity = group.group_capacity - robot.capacity

            function = Function(id_to_robots, id_to_groups)
            fault_o = 1 - function.calculate_over_load_is(id_to_robots[i])
            robot.fault_o = fault_o

    def _ini_task(self, tasks: List[Task], robots: List[Robot],
                  id_to_groups: Dict[int, Group], id_to_robots: Dict[int, Robot]):
        """Initialize task assignment to robots and groups."""
        # Assign tasks with arrive_time == -1 to groups and robots (as initial state)
        tasks_pre = []
        for task in tasks[:]:
            if task.arrive_time != -1:
                break
            else:
                tasks_pre.append(task)

        for task in tasks_pre:
            tasks.remove(task)

        # Sort tasks by size (descending) - assign largest tasks to robots with highest capacity
        tasks_pre.sort(key=lambda t: t.size, reverse=True)

        # Initialize robot-task matching
        # Priority queue: robots sorted by load/capacity ratio
        pq_robots = []
        robots.sort(key=lambda r: r.capacity, reverse=True)

        for robot in robots:
            # Update robots in group
            group_id = robot.group_id
            group = id_to_groups.get(group_id, Group())
            robot_id_in_group = group.robot_id_in_group
            if robot_id_in_group is None:
                robot_id_in_group = set()
                robot_id_in_group.add(robot.robot_id)
            else:
                robot_id_in_group.add(robot.robot_id)

            group.robot_id_in_group = robot_id_in_group
            group.group_id = group_id
            id_to_groups[group_id] = group

            self._update(tasks_pre, robot, id_to_groups)
            heapq.heappush(pq_robots, (robot.load / robot.capacity, robot.robot_id, robot))

        while tasks_pre:
            # Match all initial tasks
            _, _, robot = heapq.heappop(pq_robots)
            self._update(tasks_pre, robot, id_to_groups)
            heapq.heappush(pq_robots, (robot.load / robot.capacity, robot.robot_id, robot))

        # Fill in group capacity information
        for group_id in id_to_groups.keys():
            group = id_to_groups[group_id]
            robot_id_in_group = group.robot_id_in_group
            capacity_sum = 0.0
            for robot_id in robot_id_in_group:
                capacity_sum += id_to_robots[robot_id].capacity
            group.group_capacity = capacity_sum

    def _update(self, tasks_pre: List[Task], robot: Robot, id_to_groups: Dict[int, Group]):
        """Update robot and group with assigned task."""
        robot_tasks_list = robot.tasks_list
        if not tasks_pre:
            return

        robot_tasks_list.append(tasks_pre[0])
        # Update robot load - assign largest task to robot with smallest load
        robot.load = robot.load + tasks_pre[0].size

        group_id = robot.group_id
        # Update group load
        group = id_to_groups[group_id]
        group.group_load = group.group_load + tasks_pre[0].size
        group.group_id = group_id

        tasks_pre.pop(0)
        robot.tasks_list = robot_tasks_list

File: cascadingFailuresTaskMigration/src/main/java/main/Initialize.java
package main;

import input.Group;
import input.Robot;
import input.Task;

import java.util.*;

/***
 * 做任务迁移的起始状态下任务和智能体的匹配
 * 在级联故障的场景下，还要做故障情况（过载故障和功能故障、及未来的安全故障的初始化）
 */
public class Initialize {
    Double faultP = 0.5;
    public void run(List<Task> tasks, List<Robot> robots, HashMap<Integer, Group> idToGroups, HashMap<Integer, Robot> idToRobots) {
        iniTask(tasks, robots, idToGroups, idToRobots);
        iniFault(idToRobots,idToGroups);

    }

    private void iniFault(HashMap<Integer, Robot> idToRobots,HashMap<Integer, Group> idToGroups) {
        int size = idToRobots.size();
        int faultSize = (int) (size*faultP);
        if (faultSize==0) {
            faultSize++;
        }
        int step = size/faultSize;
        //0.1表示出现功能故障的节点占据整个系统中节点的比例
        for (int i = 0; i <size ; i++) {
            Robot robot = idToRobots.get(i);
            if (i%step==1) {
                //故障发生这块可以设置成是成片成片的，相邻的节点都发生故障（模拟级联故障的情境）
                robot.setFaultA(1);
                int groupId = robot.getGroupId();
                Group group = idToGroups.get(groupId);
                group.setGroupCapacity(group.getGroupCapacity()-robot.getCapacity());
            }
            Function function = new Function(idToRobots,idToGroups);
            Double faultO = 1-function.calculateOverLoadIS(idToRobots.get(i));
            robot.setFaultO(faultO);
            //设置组件的过载故障概率
        }
    }

    private void iniTask(List<Task> tasks, List<Robot> robots, HashMap<Integer, Group> idToGroups, HashMap<Integer, Robot> idToRobots) {
        //把时刻为-1时的任务分配给各个group及机器人（作为起始状态，只需要考虑对robot的分配即可）
        List<Task> tasksPre = new ArrayList<>();
        for (Task task : tasks) {
            if(task.getArriveTime()!=-1) {
                break;
            }else {
                tasksPre.add(task);
            }
        }
        for (Task task : tasksPre) {
            tasks.remove(task);
            //把已经分配的任务从任务待分配序列中删除
        }
        tasksPre.sort((a, b) -> Double.compare(b.getSize() ,a.getSize()));
        //从大到小排序任务大小，把最大的任务分配给Capacity越大的机器人
        //初始匹配机器人与任务
        PriorityQueue<Robot> pqRobots = new PriorityQueue<>(Comparator.comparingDouble(a -> a.getLoad() / a.getCapacity()));
        robots.sort((a,b)->Double.compare(b.getCapacity(),a.getCapacity()));
        for (Robot robot : robots) {
            //更新group里的机器人（节点编号）
            int groupId = robot.getGroupId();
            Group group = idToGroups.getOrDefault(groupId,new Group());
            Set<Integer> robotIdInGroup = group.getRobotIdInGroup();
            if (robotIdInGroup==null) {
                robotIdInGroup = new HashSet<>();
                robotIdInGroup.add(robot.getRobotId());
            } else {
                robotIdInGroup.add(robot.getRobotId());
            }
            group.setRobotIdInGroup(robotIdInGroup);
            idToGroups.put(groupId,group);

            update(tasksPre,robot,idToGroups);
            pqRobots.offer(robot);
        }
        while (!tasksPre.isEmpty()) {
            //把所有初始的任务匹配完成
            Robot robot = pqRobots.poll();
            update(tasksPre,robot,idToGroups);
            pqRobots.offer(robot);
        }
        //填上group的capacity信息
        for (Integer groupId : idToGroups.keySet()) {
            Group group = idToGroups.get(groupId);
            Set<Integer> robotIdInGroup = group.getRobotIdInGroup();
            Double capacitySum = 0.0;
            for (Integer robotId : robotIdInGroup) {
                capacitySum+=idToRobots.get(robotId).getCapacity();
            }
            group.setGroupCapacity(capacitySum);
        }
    }

    private void update(List<Task> tasksPre, Robot robot, HashMap<Integer, Group> idToGroups) {
        List<Task> robotTasksList = robot.getTasksList();
        if (tasksPre==null||tasksPre.isEmpty()) {
            return;
        }
        robotTasksList.add(tasksPre.get(0));
        //更新机器人的负载情况,把最大的任务交给负载情况最小的机器人
        robot.setLoad(robot.getLoad()+tasksPre.get(0).getSize());
        int groupId = robot.getGroupId();
        //更新group的负载
        Group group = idToGroups.get(groupId);
        group.setGroupLoad(group.getGroupLoad()+tasksPre.get(0).getSize());
        group.setGroupId(groupId);



        tasksPre.remove(0);
        robot.setTasksList(robotTasksList);
    }
}
